src/neo/SmartContract/Native/NotaryContract.cs

```diff
@@ -0,0 +1,322 @@
+            var additionalParams = (Array)data;
+            if (additionalParams.Count != 2) throw new Exception("`data` parameter should be an array of 2 elements");
+            if (!additionalParams[0].Equals(StackItem.Null)) to = additionalParams[0].GetSpan().AsSerializable<UInt160>();
+            var tx = (Transaction)engine.ScriptContainer;
+            var allowedChangeTill = tx.Sender == to;
+            var currentHeight = Ledger.CurrentIndex(engine.Snapshot);
+            Deposit deposit = GetDepositFor(engine.Snapshot, to);
+            var till = (uint)additionalParams[1].GetInteger();
+            if (till < currentHeight) throw new Exception(string.Format("`till` shouldn't be less then the chain's height {0}", currentHeight));
+            if (deposit != null && till < deposit.Till) throw new Exception(string.Format("`till` shouldn't be less then the previous value {0}", deposit.Till));
+            if (deposit is null)
+            {
+                if ((long)amount < 2 * GetNotaryServiceFeePerKey(engine.Snapshot)) throw new Exception(string.Format("first deposit can not be less then {0}, got {1}", 2 * GetNotaryServiceFeePerKey(engine.Snapshot), amount));
+                deposit = new Deposit() { Amount = 0, Till = 0 };
+                if (!allowedChangeTill) till = currentHeight + DefaultDepositDeltaTill;
+            }
+            else if (!allowedChangeTill) till = deposit.Till;
+            deposit.Amount += amount;
+            deposit.Till = till;
+            PutDepositFor(engine, to, deposit);
+        }
+
+        /// <summary>
+        /// Lock asset until the specified height is unlocked
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="addr">Account</param>
+        /// <param name="till">specified height</param>
+        /// <returns>result</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
+            if (till < Ledger.CurrentIndex(engine.Snapshot)) return false;
+            Deposit deposit = GetDepositFor(engine.Snapshot, addr);
+            if (deposit is null) return false;
+            if (till < deposit.Till) return false;
+            deposit.Till = till;
+            PutDepositFor(engine, addr, deposit);
+            return true;
+        }
+
+        /// <summary>
+        /// Withdraw sends all deposited GAS for "from" address to "to" address.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">From Account</param>
+        /// <param name="to">To Account</param>
+        /// <returns>void</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private async ContractTask Withdraw(ApplicationEngine engine, UInt160 from, UInt160 to)
```

Comment:
It should be possible to provide empty `to` address. If `to` is empty, then funds should be returned to `from`. See https://github.com/nspcc-dev/neo-go/blob/26b76ed85834a9eb96dff32ee2813060ece2f0db/pkg/core/native/notary.go#L263.

Reply:
Maybe can not provide. If the intelligent contract omits a parameter during the call, it will throw an exception, not finding the method.

Reply:
That's true, but how about providing `Null` as `to` address? That's the way how it's done in neo-go.

Reply:
Yeah, NULL is perfectly fine here, expected to be used most of the time. Although creating two methods with different number of parameters is fine too.